validator/src/aggregation/tendermint/protocol.rs

```diff
@@ -37,14 +35,60 @@ impl TendermintAggregationProtocol {
             ReplaceStore::<TendermintIdentifier, Self>::new(Arc::clone(&partitioner)),
         ));
 
+        // This will be moved into the closure.
+        let cloned_validators = Arc::clone(&validators);
+
         let evaluator = Arc::new(WeightedVote::new(
             Arc::clone(&store),
-            validators.clone(),
+            Arc::clone(&validators),
             Arc::clone(&partitioner),
-            threshold,
+            move |message: &TendermintContribution| {
+                // Conditions for final contributions:
+                // * Any proposal has 2f+1
+                // (* Nil has f+1) would be implicit in the next condition
+                // * The biggest proposal (NOT Nil) combined with the non cast votes is less than 2f+1
+
+                let mut uncast_votes = policy::Policy::SLOTS as usize;
+                let mut biggest_weight = 0;
+
+                // Check every proposal present, including None
+                for (hash, signature) in message.contributions.iter() {
+                    let weight = cloned_validators
+                        .signers_weight(&signature.signers)
+                        .unwrap_or(0);
+                    // Any option which has at least 2f+1 votes make this contribution final.
+                    if weight >= policy::Policy::TWO_F_PLUS_ONE as usize {
+                        return true;
+                    }
+                    // If the proposal does not have 2f+1 check if it has the new biggest_weight
+                    if hash.is_some() && weight > biggest_weight {
+                        biggest_weight = weight;
+                    }
+                    // The cast votes need to be removed from the uncast votes
+                    uncast_votes -= weight;
+                }
+
+                let cast_votes = policy::Policy::SLOTS as usize - uncast_votes;
```

Comment:
Couldn't we better use `saturating_sub`?

Reply:
Here it is not strictly necessary, as `uncast_votes` start as `policy::Policy::SLOTS` and then only ever decrease. With the `saturating_sub` from your previous comment it can no longer underflow and thus can only ever reach 0. I still added it, because it does no harm at all and is a better practice.
Thanks.